lib/message-handler.js

```diff
@@ -177,539 +183,545 @@ class MessageHandler {
-                                options.raw = Buffer.concat(options.rawchunks, options.chunklen);
-                            } else {
-                                options.raw = Buffer.concat(options.rawchunks);
-                            }
-                        }
-                        // user has encryption on or target mailbox encrypted, encrypt message and prepare again
-                        // do not encrypt drafts
-                        // may have a situation where we got prepared and no options.raw but options.rawchunks instead, concat them
-                        this.encryptMessage(userData.pubKey, options.raw, (err, res) => {
-                            if (err) {
-                                return callback(err);
-                            }
+        let thread;
 
-                            if (res) {
-                                // new encrypted raw available
-                                options.raw = res;
-                            }
+        try {
+            thread = await this.getThreadIdAsync(mailboxData.user, subject, mimeTree);
+        } catch (err) {
+            return rollback(err);
+        }
 
-                            delete options.prepared; // delete any existing prepared as new will be generated
-                            this.prepareMessage(options, (err, newPrepared) => {
-                                if (err) {
-                                    return callback(err);
+        messageData.thread = thread;
+
+        let insertRes;
+
+        try {
+            insertRes = await this.database.collection('messages').insertOne(messageData, { writeConcern: 'majority' });
+        } catch (err) {
+            return rollback(err);
+        }
+
+        if (!insertRes || !insertRes.acknowledged) {
+            let err = new Error('Failed to store message [1]');
+            err.responseCode = 500;
+            err.code = 'StoreError';
+            return rollback(err);
+        }
+
+        let logTime = messageData.meta.time || new Date();
+        if (typeof logTime === 'number') {
+            logTime = new Date(logTime);
+        }
+
+        let uidValidity = mailboxData.uidValidity;
+        let uid = messageData.uid;
+
+        const finishFunc = // finishFunc:
+            async () =>
+                new Promise(resolve => {
+                    this.notifier.addEntries(
+                        mailboxData,
+                        {
+                            command: 'EXISTS',
+                            uid: messageData.uid,
+                            ignore: options.session && options.session.id,
+                            message: messageData._id,
+                            modseq: messageData.modseq,
+                            unseen: messageData.unseen,
+                            idate: messageData.idate,
+                            thread: messageData.thread
+                        },
+                        async () => {
```

Comment:
addEntries callback can not be async, must be regular callback function. If it throws then will not be caught